File: bench-adapters/python/benches/jaro_bench.py
"""Head-to-head: StringCheese (via wasm) vs. Python's Jaro / Jaro-Winkler libs.

# Contestants

Jaro:

* **StringCheese** — ``jaro_similarity`` from the wasm component.
* **python-Levenshtein.jaro** — C extension.
* **jellyfish.jaro_similarity** — Rust extension.

Jaro–Winkler:

* **StringCheese** — ``jaro_winkler_similarity`` (classic: prefix 4,
  scaling 0.1, no boost threshold).
* **python-Levenshtein.jaro_winkler** — same defaults as the classic
  Winkler paper.
* **jellyfish.jaro_winkler_similarity** — same defaults.

rapidfuzz **is not** included in this file. ``rapidfuzz.distance.Jaro``
and ``rapidfuzz.distance.JaroWinkler`` exist and would slot in cleanly;
the ecosystem-standard-pair for Jaro-family similarity in Python is
python-Levenshtein + jellyfish, and adding rapidfuzz is a followup.

# Variant identity

All three libraries here compute **classic** Jaro–Winkler with prefix 4
and scaling 0.1. StringCheese's ``JaroWinkler::classic()`` matches those
defaults; ``docs/design/`` and ``component/README.md`` are clear that
only the classic variant is exposed at the WIT boundary. If a future
non-classic tuning goes into the WIT interface, that pairing needs a
separate bench file to keep the head-to-head axis clean.
"""

from __future__ import annotations

import Levenshtein
import jellyfish
import pytest

from _inputs import LENGTHS, REGIMES, build_pair

_SALTS = (0xD1, 0xD2, 0xD3)


@pytest.fixture(scope="module")
def pairs() -> dict[tuple[int, str], tuple[bytes, bytes]]:
    return {(n, k): build_pair(n, k, _SALTS) for n in LENGTHS for k in REGIMES}


# --------------------------------------------------------------------------- #
# Jaro
# --------------------------------------------------------------------------- #


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_stringcheese_jaro(benchmark, stringcheese, pairs, n, kind):
    benchmark.group = f"jaro/{kind}/len{n:04d}"
    a, b = pairs[(n, kind)]
    benchmark(stringcheese.jaro_similarity, a, b)


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_python_levenshtein_jaro(benchmark, pairs, n, kind):
    benchmark.group = f"jaro/{kind}/len{n:04d}"
    a_b, b_b = pairs[(n, kind)]
    a, b = a_b.decode("ascii"), b_b.decode("ascii")
    benchmark(Levenshtein.jaro, a, b)


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_jellyfish_jaro(benchmark, pairs, n, kind):
    benchmark.group = f"jaro/{kind}/len{n:04d}"
    a_b, b_b = pairs[(n, kind)]
    a, b = a_b.decode("ascii"), b_b.decode("ascii")
    benchmark(jellyfish.jaro_similarity, a, b)


# --------------------------------------------------------------------------- #
# Jaro–Winkler
# --------------------------------------------------------------------------- #


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_stringcheese_jaro_winkler(benchmark, stringcheese, pairs, n, kind):
    benchmark.group = f"jaro_winkler/{kind}/len{n:04d}"
    a, b = pairs[(n, kind)]
    benchmark(stringcheese.jaro_winkler_similarity, a, b)


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_python_levenshtein_jaro_winkler(benchmark, pairs, n, kind):
    benchmark.group = f"jaro_winkler/{kind}/len{n:04d}"
    a_b, b_b = pairs[(n, kind)]
    a, b = a_b.decode("ascii"), b_b.decode("ascii")
    benchmark(Levenshtein.jaro_winkler, a, b)


@pytest.mark.parametrize("n", LENGTHS, ids=[f"len{n:04d}" for n in LENGTHS])
@pytest.mark.parametrize("kind", REGIMES)
def test_jellyfish_jaro_winkler(benchmark, pairs, n, kind):
    benchmark.group = f"jaro_winkler/{kind}/len{n:04d}"
    a_b, b_b = pairs[(n, kind)]
    a, b = a_b.decode("ascii"), b_b.decode("ascii")
    benchmark(jellyfish.jaro_winkler_similarity, a, b)

File: bench-adapters/js/benches/jaro.bench.js
// Head-to-head: StringCheese (via wasm) vs. JS Jaro / Jaro–Winkler libs.
//
// Contestants
// ===========
//
// Jaro-Winkler:
//
// * **StringCheese** — `jaroWinklerSimilarity` (classic: prefix 4,
//   scaling 0.1, no boost threshold).
// * **natural.JaroWinklerDistance** — the same classic defaults; the
//   `natural` package's function name says "Distance" but it actually
//   returns the Jaro–Winkler *similarity* in `[0.0, 1.0]` (see
//   `node_modules/natural/lib/natural/distance/jaro-winkler_distance.js`
//   — it returns `jaro + l * p * (1 - jaro)`, i.e. a similarity).
//
// Jaro (pure):
//
// * **StringCheese** — `jaroSimilarity`.
// * **Ecosystem gap:** `natural` does not expose the pure Jaro
//   similarity as a top-level function — its Jaro implementation is a
//   file-scope `distance()` inside `jaro-winkler_distance.js`. Neither
//   `fastest-levenshtein`, `js-levenshtein`, nor `string-similarity`
//   expose Jaro at all. The StringCheese cell is therefore benched
//   *alongside* a `string-similarity.compareTwoStrings` reference —
//   documented as "not directly comparable" because compareTwoStrings
//   computes a Dice coefficient over bigrams, not a Jaro similarity.
//   The point is to make the FFI-vs-native-JS cost visible at each
//   length; the numeric result is not comparable.
//
// Representation caveat
// =====================
//
// Same as the other files: StringCheese consumes `Uint8Array`, the
// native libraries consume `string`. On ASCII input the semantics
// agree; the FFI cost is folded in on purpose.

// `natural` transitively pulls in `dotenvx`; see levenshtein.bench.js
// for the ESM-hoisting rationale behind the side-effect import.
import "./_env.js";

import natural from "natural";
import * as stringSimilarity from "string-similarity";

import { StringCheese } from "../stringcheese_adapter.js";
import { LENGTHS, REGIMES, buildPair, lenLabel } from "./_inputs.js";
import { makeBench, runAll } from "./_runner.js";

const { JaroWinklerDistance: NaturalJW } = natural;

// Shared with the Rust and Python adapters — same salts because the
// Jaro match-window behaviour is best observed against a known-shared
// corpus for cross-adapter reasoning.
const SALTS = /** @type {[number, number, number]} */ ([0xd1, 0xd2, 0xd3]);

async function main() {
  const sc = new StringCheese();
  const decoder = new TextDecoder("ascii");

  /** @type {Map<string, {aB: Uint8Array, bB: Uint8Array, aS: string, bS: string}>} */
  const pairs = new Map();
  for (const n of LENGTHS) {
    for (const kind of REGIMES) {
      const [aB, bB] = buildPair(n, kind, SALTS);
      pairs.set(`${kind}/${lenLabel(n)}`, {
        aB,
        bB,
        aS: decoder.decode(aB),
        bS: decoder.decode(bB),
      });
    }
  }

  const benches = [];

  // Pure Jaro. StringCheese is the only turnkey Jaro in the pure-JS
  // ecosystem; string-similarity's `compareTwoStrings` is included as
  // a "not directly comparable" FFI-cost anchor. See the file docstring.
  for (const kind of REGIMES) {
    for (const n of LENGTHS) {
      const key = `${kind}/${lenLabel(n)}`;
      const { aB, bB, aS, bS } = pairs.get(key);
      const bench = makeBench(`jaro/${key}`);
      bench.add("stringcheese", () => sc.jaroSimilarity(aB, bB));
      bench.add("string-similarity (NOT Jaro)", () =>
        stringSimilarity.compareTwoStrings(aS, bS),
      );
      benches.push(bench);
    }
  }

  // Jaro-Winkler. StringCheese vs. natural — both compute the classic
  // Winkler variant with prefix 4, scaling 0.1, no boost threshold.
  for (const kind of REGIMES) {
    for (const n of LENGTHS) {
      const key = `${kind}/${lenLabel(n)}`;
      const { aB, bB, aS, bS } = pairs.get(key);
      const bench = makeBench(`jaro_winkler/${key}`);
      bench.add("stringcheese", () => sc.jaroWinklerSimilarity(aB, bB));
      bench.add("natural", () => NaturalJW(aS, bS));
      benches.push(bench);
    }
  }

  await runAll(benches);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
